helm-info.el

```diff
@@ -240,15 +241,71 @@ helm-info-<CANDIDATE>."
       (funcall helm-info-function)
       (ring-insert helm-info-searched candidate))))
 
+(defvar helm-info--files-cache nil)
+(defvar helm-info--files-doc-cache (make-hash-table :test 'equal))
+
+(defun helm-info-file-doc (file)
+  "Return dir entry from the Info FILE."
+  (or
+   (with-temp-buffer
+     (ignore-errors (info-insert-file-contents file))
+     (goto-char (point-min))
+     (when (re-search-forward (rx line-start "START-INFO-DIR-ENTRY" line-end)
+                              nil t)
+       (forward-line 1)
+       (when (re-search-forward
+              (rx line-start
+                  "*" (one-or-more whitespace)
+                  (group (one-or-more (not ":")))
+                  ":" (one-or-more whitespace)
+                  "(" (one-or-more (not ")"))
+                  ")." (one-or-more whitespace)
+                  (group (one-or-more any)))
+              (pos-eol) t)
+         (format "%s: %s"
+                 (match-string 1) (match-string 2)))))
+   "No summary"))
+
 (defun helm-def-source--info-files ()
   "Return a Helm source for Info files."
+  (require 'helm-mode)
   (helm-build-sync-source "Helm Info"
     :candidates
     (lambda () (copy-sequence helm-default-info-index-list))
     :candidate-number-limit 999
     :candidate-transformer
     (lambda (candidates)
-      (sort candidates #'string-lessp))
+      (let ((candidates (sort candidates #'string-lessp))
+            (longest (apply #'max (mapcar #'length candidates)))
+            done)
+        (cl-loop with reporter = (unless done
+                                   (make-progress-reporter
+                                    "Scanning libraries..." 0 (length candidates)))
+                 for c in candidates
+                 for count from 0
+                 for sep = (helm-make-separator c longest)
+                 for file = (or
+                             (assoc-default c helm-info--files-cache)
+                             (let ((file (Info-find-file c)))
+                               (push (cons c file) helm-info--files-cache)
+                               file))
```

Comment:
When `Info-find-file` fails to find a file it popup the top dir info buffer as default, which override the helm buffer, I fixed it like this:

```elisp
(let ((file (Info-find-file c t)))
   (when file (push (cons c file) helm-info--files-cache))
   file)
```
Use the noerror arg of Info-find-file.